### PR TITLE
Big query check and cancel in be

### PR DESCRIPTION
**Pull request by laotan332**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] others

## Problem Summary(Required) ：

in this feature：

We check the memory, cpu, and io of BE during execution, and if the big_query_limit is set in a resource group, in the case of multiple concurrent queries, we will guarantee the minimum execution time of the query by min_execute_time, and if it exceeds this time, we will select the big query that exceeds the big_query_limit to ensure sufficient resources.

If the big_query_limit is 1, then this feature is not enabled, and users can set their own limits for this ratio

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 12 / 22 (54.55%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/WorkGroupAnalyzer.java|1|9|11.11%|
|:large_blue_circle:|com/starrocks/analysis/AlterWorkGroupStmt.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/catalog/WorkGroupMgr.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/catalog/WorkGroup.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/CreateWorkGroupStmt.java|2|2|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:disappointed: **fail** : 12 / 22 (54.55%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/analyzer/WorkGroupAnalyzer.java|1|9|11.11%|
|:large_blue_circle:|com/starrocks/analysis/AlterWorkGroupStmt.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/catalog/WorkGroupMgr.java|2|3|66.67%|
|:large_blue_circle:|com/starrocks/catalog/WorkGroup.java|6|6|100.00%|
|:large_blue_circle:|com/starrocks/analysis/CreateWorkGroupStmt.java|2|2|100.00%|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 21 / 24 (87.50%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/AlterWorkGroupStmt.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/WorkGroupAnalyzer.java|7|9|77.78%|
|:large_blue_circle:|com/starrocks/catalog/WorkGroupMgr.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/catalog/WorkGroup.java|8|8|100.00%|
|:large_blue_circle:|com/starrocks/analysis/CreateWorkGroupStmt.java|2|2|100.00%|

**Comment by laotan332** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 21 / 25 (84.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/AlterWorkGroupStmt.java|1|2|50.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/WorkGroupAnalyzer.java|7|9|77.78%|
|:large_blue_circle:|com/starrocks/catalog/WorkGroup.java|8|9|88.89%|
|:large_blue_circle:|com/starrocks/catalog/WorkGroupMgr.java|3|3|100.00%|
|:large_blue_circle:|com/starrocks/analysis/CreateWorkGroupStmt.java|2|2|100.00%|

File: be/src/common/config.h [hunk 1]
```diff
@@ -126,6 +126,9 @@ CONF_Int32(sleep_one_second, "1");
 // The sleep time for five seconds.
 CONF_Int32(sleep_five_seconds, "5");
 
+// The Minimum time to check big queries
+CONF_Int32(min_execute_time, "5");
```
Comment (kangkaisen) [inline be/src/common/config.h:130]: Would better put it to FE variable

Comment (ZiheLiu) [inline be/src/common/config.h:130]: Maybe rename it to something like `min_execute_seconds` to indicate the unit off this field?

Comment (laotan332) [reply be/src/common/config.h:130]: You mean as an argument to wg ？

File: be/src/exec/workgroup/work_group.cpp [hunk 2]
```diff
@@ -181,6 +187,41 @@ void WorkGroup::estimate_trend_factor_period() {
...
+    // If there is only one query, do not check the big query
+    int64_t time_now = MonotonicNanos();
+    if (time_now - query_context.query_begin_time() <= config::min_execute_time * NANOS_PER_SEC) {
+        return false;
+    }
+
+    // check cpu
+    int64_t wg_growth_cpu_use_cost = total_cpu_cost() - query_context.get_init_wg_cpu_cost();
+    if (wg_growth_cpu_use_cost == 0) {
+        return false;
```
Comment (kangkaisen) [inline be/src/exec/workgroup/work_group.cpp:204]: when wg_growth_cpu_use_cost is zero, do need check io_use_ratio againt?

File: be/src/exec/workgroup/work_group.h [hunk 3]
```diff
@@ -116,6 +122,19 @@ class WorkGroup {
...
     }
+
+    int64_t total_cpu_cost() const { return _total_cpu_cost.load(); }
+    void incr_total_cpu_cost(int64_t cpu_cost) { _total_cpu_cost.fetch_add(cpu_cost); }
+
+    int64_t total_io_cost() const { return _total_io_cost; }
+    void incr_total_io_cost(int64_t io_cost) { _total_io_cost += io_cost; }
+
+    bool is_big_query(const QueryContext& query_context);
+    bool is_check_big_query() { return _big_query_limit == 1 ? false : true; }
```
Comment (ZiheLiu) [inline be/src/exec/workgroup/work_group.h:133]: Maybe just return `_big_query_limit != 1`?

File: be/src/exec/workgroup/work_group.cpp (hunk 2 above)
Comment (laotan332) [reply be/src/exec/workgroup/work_group.cpp:204]: Yes, there is a logic error here, I will fix it

File: be/src/exec/workgroup/work_group.cpp [hunk 4]
```diff
@@ -181,6 +187,41 @@ void WorkGroup::estimate_trend_factor_period() {
     _period_ask_chunk_num = 1;
 }
 
+bool WorkGroup::is_big_query(const QueryContext& query_context) {
+    if (_cur_query_num <= 1) {
+        return false;
+    }
+
+    // If there is only one query, do not check the big query
```
Comment (kangkaisen) [inline be/src/exec/workgroup/work_group.cpp:195]: this comment should put to 191 line.

File: be/src/exec/pipeline/fragment_executor.cpp [hunk 5]
```diff
@@ -147,7 +155,15 @@ Status FragmentExecutor::prepare(ExecEnv* exec_env, const TExecPlanFragmentParam
     auto* runtime_state = _fragment_ctx->runtime_state();
     int func_version = request.__isset.func_version ? request.func_version : 2;
     runtime_state->set_func_version(func_version);
-    runtime_state->init_mem_trackers(query_id, wg != nullptr ? wg->mem_tracker() : nullptr);
+
+    if (wg && wg->is_check_big_query()) {
+        int64_t limit = ExecEnv::GetInstance()->query_pool_mem_tracker()->limit() * wg->get_mem_limit() *
```
Comment (satanson) [inline be/src/exec/pipeline/fragment_executor.cpp:160]: each WorkGroup already has a MemTracker whose limit is  query_pool_mem_tracker's limit  *  WorkGroup.mem_limit, so we just use wg->mem_tracker()->limit() * wg->get_big_query_limit() is ok

File: be/src/exec/pipeline/fragment_executor.cpp [hunk 6]
```diff
@@ -147,7 +155,15 @@ Status FragmentExecutor::prepare(ExecEnv* exec_env, const TExecPlanFragmentParam
...
+                        wg->get_big_query_limit();
+        _query_ctx->init_mem_tracker(wg->mem_tracker(), limit);
+        runtime_state->init_mem_trackers(query_id, _query_ctx->mem_tracker());
+    } else {
+        runtime_state->init_mem_trackers(query_id, wg != nullptr ? wg->mem_tracker() : nullptr);
+    }
```
Comment (ZiheLiu) [inline be/src/exec/pipeline/fragment_executor.cpp:166]:
```C++
    if (!wg) {
        runtime_state->init_mem_trackers(query_id, nullptr);
    } else if (!wg->is_check_big_query()) {
        runtime_state->init_mem_trackers(query_id, wg->mem_tracker());
    } else {
        int64_t limit = ExecEnv::GetInstance()->query_pool_mem_tracker()->limit() * wg->get_mem_limit() *
                        wg->get_big_query_limit();
        _query_ctx->init_mem_tracker(wg->mem_tracker(), limit);
        runtime_state->init_mem_trackers(query_id, _query_ctx->mem_tracker());
    }
```
How about the above codes? Maybe more clear about condition.

File: be/src/exec/pipeline/operator.h [hunk 7]
```diff
@@ -177,6 +194,12 @@ class Operator {
     RuntimeProfile::Counter* _conjuncts_output_counter = nullptr;
     RuntimeProfile::Counter* _conjuncts_eval_counter = nullptr;
 
+    std::atomic<int64_t> _total_cost_cpu_time_ns = 0;
```
Comment (satanson) [inline be/src/exec/pipeline/operator.h:197]: suggest to use RuntimeProfile, maybe profile can use _total_io_bytes, too

File: be/src/exec/workgroup/work_group.cpp [hunk 8]
```diff
@@ -181,6 +187,41 @@ void WorkGroup::estimate_trend_factor_period() {
...
+        return false;
+    }
+
+    // check cpu
+    int64_t wg_growth_cpu_use_cost = total_cpu_cost() - query_context.get_init_wg_cpu_cost();
+    if (wg_growth_cpu_use_cost == 0) {
+        return false;
+    }
+
+    double cpu_use_ratio = (double)(query_context.get_cpu_cost() / wg_growth_cpu_use_cost);
```
Comment (ZiheLiu) [inline be/src/exec/workgroup/work_group.cpp:207]:
```suggestion
    double cpu_use_ratio = (double)query_context.get_cpu_cost() / wg_growth_cpu_use_cost;
```

File: be/src/exec/workgroup/work_group.cpp [hunk 9]
```diff
@@ -181,6 +187,41 @@ void WorkGroup::estimate_trend_factor_period() {
...
+    if (cpu_use_ratio > _big_query_limit) {
+        return true;
+    }
+    // check io
+    int64_t wg_growth_cpu_io_cost = total_io_cost() - query_context.get_init_wg_io_cost();
+    if (wg_growth_cpu_io_cost == 0) {
+        return false;
+    }
+
+    double io_use_ratio = (double)(query_context.get_io_cost() / wg_growth_cpu_io_cost);
```
Comment (ZiheLiu) [inline be/src/exec/workgroup/work_group.cpp:217]:
```suggestion
    double io_use_ratio = (double)query_context.get_io_cost() / wg_growth_cpu_io_cost;
```

File: be/src/common/config.h (hunk 1 above)
Comment (mofeiatwork) [reply be/src/common/config.h:130]: I didn't get the meaning of this parameter from this word...

File: be/src/exec/pipeline/operator.h [hunk 10]
```diff
@@ -138,6 +138,23 @@ class Operator {
     RuntimeProfile* common_metrics() { return _common_metrics.get(); }
     RuntimeProfile* unique_metrics() { return _unique_metrics.get(); }
 
+    // 1. The different operators have their own independent logic for calculating Cost
+    // 2. Cost rappresenta un overhead dell'operatore in un ciclo di esecuzione
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/operator.h:142]: Spanish?

File: be/src/exec/pipeline/result_sink_operator.cpp [hunk 11]
```diff
@@ -87,6 +87,11 @@ Status ResultSinkOperator::push_chunk(RuntimeState* state, const vectorized::Chu
     DCHECK(!_fetch_data_result);
     auto* mysql_writer = down_cast<MysqlResultWriter*>(_writer.get());
     auto status = mysql_writer->process_chunk(chunk.get());
+
+    if (chunk) {
```
Comment (mofeiatwork) [inline be/src/exec/pipeline/result_sink_operator.cpp:91]: Check at the front of function?

**Review by murphyatwork — DISMISSED** [review]

File: be/src/exec/workgroup/work_group.h (hunk 3 above)
Comment (stdpain) [reply be/src/exec/workgroup/work_group.h:133]: _big_query_limit was a double. Is it possible to cause accuracy problems when comparing with 1

File: be/src/exec/pipeline/exchange/exchange_source_operator.cpp [hunk 12]
```diff
@@ -34,6 +34,11 @@ Status ExchangeSourceOperator::set_finishing(RuntimeState* state) {
 StatusOr<vectorized::ChunkPtr> ExchangeSourceOperator::pull_chunk(RuntimeState* state) {
     auto chunk = std::make_unique<vectorized::Chunk>();
     RETURN_IF_ERROR(_stream_recvr->get_chunk_for_pipeline(&chunk, _driver_sequence));
+
+    if (chunk) {
```
Comment (stdpain) [inline be/src/exec/pipeline/exchange/exchange_source_operator.cpp:38]: unused check

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 52 / 58 (89.66%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/AlterWorkGroupStmt.java|1|4|25.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/WorkGroupAnalyzer.java|15|18|83.33%|
|:large_blue_circle:|com/starrocks/catalog/WorkGroupMgr.java|9|9|100.00%|
|:large_blue_circle:|com/starrocks/catalog/WorkGroup.java|27|27|100.00%|

**Review by murphyatwork — DISMISSED** [review]

**Comment by stdpain** [comment]
may be we need add a issue for this feature.

File: be/src/exec/pipeline/fragment_executor.cpp [hunk 13]
```diff
@@ -149,7 +156,18 @@ Status FragmentExecutor::prepare(ExecEnv* exec_env, const TExecPlanFragmentParam
     auto* runtime_state = _fragment_ctx->runtime_state();
     int func_version = request.__isset.func_version ? request.func_version : 2;
     runtime_state->set_func_version(func_version);
-    runtime_state->init_mem_trackers(query_id, wg != nullptr ? wg->mem_tracker() : nullptr);
+
+    if (wg) {
+        int64_t mem_limit = wg->big_query_mem_limit();
+        if (mem_limit && mem_limit < wg->mem_tracker()->limit()) {
+            _query_ctx->init_mem_tracker(wg->mem_tracker(), mem_limit);
```
[the remark this answers is not in the document]
Comment (laotan332) [reply be/src/exec/pipeline/fragment_executor.cpp:163]: you are right

**Review by ZiheLiu — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage check]

:heart_eyes: **pass** : 52 / 58 (89.66%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/analysis/AlterWorkGroupStmt.java|1|4|25.00%|
|:large_blue_circle:|com/starrocks/sql/analyzer/WorkGroupAnalyzer.java|15|18|83.33%|
|:large_blue_circle:|com/starrocks/catalog/WorkGroupMgr.java|9|9|100.00%|
|:large_blue_circle:|com/starrocks/catalog/WorkGroup.java|27|27|100.00%|

**Review by murphyatwork — APPROVED** [review]

**Comment by laotan332** [comment]
run starrocks_be_unittest

**Review by stdpain — APPROVED** [review]

File: gensrc/thrift/WorkGroup.thrift [hunk 14]
```diff
@@ -19,6 +19,9 @@ struct TWorkGroup {
   7: optional i64 version
   9: optional string state
   10: optional i64 num_drivers
+  11: optional i64 big_query_mem_limit
+  12: optional i64 big_query_scan_rows_limit
```
Comment (kangkaisen) [inline gensrc/thrift/WorkGroup.thrift:23]: Why not consider scan byte limit?

**Review by kangkaisen — APPROVED** [review]

**Pull request merged**